examples/bri-3/src/bri/identity/bpiSubjects/capabilities/updateBpiSubject/updateBpiSubjectCommand.handler.ts

```diff
@@ -1,23 +1,19 @@
 import { CommandHandler, ICommandHandler } from "@nestjs/cqrs";
 import { BpiSubjectAgent } from "../../agents/bpiSubjects.agent";
+import { BpiSubjectRepository } from "../../persistence/bpiSubjects.repository";
 import { UpdateBpiSubjectCommand } from "./updateBpiSubject.command";
 
 @CommandHandler(UpdateBpiSubjectCommand)
 export class UpdateBpiSubjectCommandHandler implements ICommandHandler<UpdateBpiSubjectCommand> {
-  constructor(private agent: BpiSubjectAgent) {}
+  constructor(private agent: BpiSubjectAgent, private repo: BpiSubjectRepository) {}
 
   async execute(command: UpdateBpiSubjectCommand) {
-    const { id, name, desc, publicKey } = command;
+    const bpiSubjectToUpdate = await this.agent.fetchUpdateCandidateAndThrowIfUpdateValidationFails(command.id, command.name, command.desc, command.publicKey);
 
-    // this.agent.throwIfUpdateBpiSubjectInvalid(name, desc, publicKey);
+    this.agent.updateBpiSubject(bpiSubjectToUpdate, command.name, command.desc, command.publicKey);
 
-    // const bpiSubjectToUpdate = this.agent.fetchBpiSubjectToUpdate(id);
-    // this.agent.updateBpiSubject(bpiSubjectToUpdate, name, desc, publicKey);
-
-    // TODO: Generic map of domain model to entity model
-    // this.orm.store(newBpiSubject);
-
-    // TODO: Response DTO
-    return true;
+    await this.repo.updateBpiSubject(bpiSubjectToUpdate);
```

Comment:
just for discussion purposes it seems that there is no clear line if agent is doing db calls, or just validation etc
for example you could write this method as 

this.repo.getSubject()
this.agent.validateSubject()
this.repo.updateSubject()

or

this.agent.getSubjectAndValidate();
this.agentUpdate() which would also call repo and you can call method updateBpiSubjectAndThrowIfItFails ...

or couple other combinations like implementation in PR

dont have hard opinion on this, but curious on your thoughts about clear responsibilities

Reply:
Very good point. Renamed repo to agent. Explanation bellow:

We are early in the implementation so we do not have enough good examples in code but a good way to think about it is the following: 

Agent represents a person in an organization, or even better a role. 

Think about a hotel:

Each role is tasked with specific responsibilities. Lift boy will operate the elevator, receptionists will  welcome the guests and serve them at the reception...

In case that a person in a role is overloaded by too much work and\or too much responsibilities, organization would introduce additional persons and potentially additional roles, to help solve the problem. Think of it as horizontal and vertical scaling. Receptionist could have his\her role relieved by introducing another role (i.e. Cashier) that deals only with payments from the guests, or by introducing another receptionist.  

In our case, we currently have bpiSubjectAgent and bpiSubjectStorageAgent. One deals with everything BpiSubject related apart from storage - which is the job of the storage agent. As we grow, we might decide to split the bpiSubjectAgent into multiple different agents (new roles) which would help us keep the amount of responsibilities in a role on a level that is easy to read and maintain. As the user load grows, we might decide to introduce a  pub\sub bus were multiple agents of the same kind (new persons) listen to events and process them (microservices).

A command handler can inject multiple different agents in order to orchestrate the execution of  the command at hand. It is tasked with declaring WHAT needs to happen (thus agent method names should be highly descriptive) and in what order.

The agents are tasked with HOW. Agents can share among themselves helper services they might need (think of them as tools - both cashier and receptionist rely on the same Excel sheet to give the guest a preliminary price calc).

Because of the fact that you could have multiple agents in a single command who are all part of a single business transaction, we need to treat the storage role as an agent so that we can commit transaction as a last step of the execution of the command. Otherwise storage could be abstracted away behind an agent.

Reply:
> Because of the fact that you could have multiple agents in a single command who are all part of a single business 
> transaction, we need to treat the storage role as an agent so that we can commit transaction as a last step of the 
> execution of the command. Otherwise storage could be abstracted away behind an agent.

@ognjenkurtic each agent in a command might need storage as a tool. so should we have storage as BOTH agent and a helper service I wonder? cc @skosito

Reply:
I think of tools usually as pure functions - they do not produce side effects. In our domain, a proper tool is for example a class (service) that is responsible for performing signature verification. Input from the caller -> Verification -> Output to the caller - all without changing the global state of the application. When thinking about storage in this light it is clear that it is not a tool - it deals with a cross-cutting concern (DB transaction). 

If all agents are allowed to perform storage, you as a dev would need to actually read the implementation of every agent method used in the command to make sure that the result of the business transaction is properly stored or rejected. This would quickly turn into a mess to maintain and extend as we develop further.

If, on the other hand, only the storageAgent performs the commit of the transaction at the end of the command execution, this tracking of the transaction becomes trivial. The command handler which describes the WHAT and the order of execution becomes the only entity that is responsible for accepting or rejecting the storage of the result of of work of all the previous agents, based on the  validity of the business transaction.

If there is a need to execute multiple DB transactions as part of a single API call, this is usually a sign that you should split the command into multiple commands and chain them. This is what is called Saga in NestJS. 

I suggest we continue with this approach - as we get to more complex business cases, the usage and the motivation behind the architecture will become more clear. We can always adjust and refactor as we go, if the need arises.